File: 백준/Gold/1106. 호텔/호텔.js
const fs = require("fs");
const filePath = process.platform == "linux" ? "/dev/stdin" : "input.txt";
const input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split("\n")
  .map((line) => line.split(" ").map((n) => Number(n)));

const [C, N] = input.shift();
const maxNum = Math.max(...input.map((e) => e[1]));
const maxIdx = C + maxNum;

const dp = Array(maxIdx + 1).fill(100000); // idx명 유치 최소 비용

for (const [a, b] of input) {
  dp[b] = Math.min(dp[b], a);
}

for (const [cost, customers] of input) {
  for (let i = customers; i < maxIdx; ++i) {
    dp[i] = Math.min(dp[i], dp[i - customers] + cost);
  }
}

console.log(Math.min(...dp.slice(C)));

File: 백준/Gold/1106. 호텔/호텔.cc
#include <iostream>
using namespace std;

int MAX = 100000;
int main() {
  int c, n;
  cin >> c >> n;
  int dp[c+1] = {0, }; // dp[i]: i명 유치하는데 드는 최소비용
  for (int i=1; i<=c; ++i) dp[i]=MAX;
  int a, b;  // 비용, 고객의 수
  for (int _=0; _<n; ++_) {
    cin >> a >> b;
    for (int cnt=0; cnt<=c; cnt++) {
      int i = cnt+b > c ? c : cnt+b; // 최대 인덱스 제한
      dp[i]=min(dp[i], dp[cnt]+a);
    }
  }

  cout << dp[c];
}